CI: fix bash complaining about "unexpected tokens" (parenthesis)

qmonnet opened the pull request:
In #11094, @pchaigno reported that a few K8sServices tests contain parenthesis in their names, and that this creates confusion for the bash command used to create the bugtool archive directory if one of those tests fails on the CI.

Let's fix this by enquoting the directory name in the relevant bash commands, to avoid the parenthesis to be interpreted. This was tested in draft PR #11230, where test `Tests NodePort (kube-proxy)` is forced to fail and where we can grep the CI logs for the `unexpected token` bash error.

While grepping for that message, it appeared that the logs had other occurrences of this `unexpected token` error. They are not related to the above issues, but instead come from a wrong usage of the optional arguments for `helpers.CurlFail()` at one location in K8sServices tests. Fix this one by removing the spurious (unnecessary) argument. The aforementioned PR also has this fix, and shows that it solves the complaint from bash.

Fixes: #11094

nebril reviewed (commented):
Good catch! LGTM overall, one question left inline, let's merge it when we are it's not breaking CI.

nebril commented on test/k8sT/Services.go:
```diff
@@ -385,7 +385,7 @@ var _ = Describe("K8sServicesTest", func() {
 			By("Making %d curl requests from %s to %q", count, fromPod, url)
 			for i := 1; i <= count; i++ {
-				res, err := kubectl.ExecInHostNetNS(context.TODO(), fromPod, helpers.CurlFail(url, "--max-time 3"))
+				res, err := kubectl.ExecInHostNetNS(context.TODO(), fromPod, helpers.CurlFail(url))
```
Why is this changed? Can this be cause of our test fail?

qmonnet replied:
From the commit log:
```
helpers.CurlFail() can take optional arguments, and at one location this
is used to pass the "--max-time" option to the curl command line.

But these optional arguments for helpers.CurlFail() are not intended to
be used that way; they should instead be used to format the target URL,
which is supposed to be a string containing the relevant format
specifiers. Because in the current case, the URL provided has no format
specifier, go appends the additional arguments at the end, with
extraneous characters. So instead of producing the final curl command:

   curl [...] http://192.168.36.11:30861 --max-time 3 [...]

We actually get:

   curl [...] --max-time 8 http://192.168.36.11:30861%!(EXTRA string=--max-time 3)

The first "--max-time 8" is added by helpers.CurlFail() itself.

Then of course bash fails to interpret that command, and complains with:

   bash: -c: line 0: syntax error near unexpected token `('

In order to fix it, let's remove the "--max-time 3" argument entirely.
As we can see from the resulting command line, "--max-time" is already
used (with "CurlMaxTimeout") anyway. We will wait a few seconds more,
but at least the curl command will effectively be run.
```

So with that change, the `curl` command _is_ run (previously bash would fail to parse the command and would not run it), and this apparently caused the test to fail (`curl` passed when it was supposed to fail). But this only means that the test was not run correctly so far. So if the `curl` command runs and fail, I understand that either the test is incorrect, or it raises a bug in Cilium. I haven't found time to investigate further for now.

gandro replied:
Apologies for the late reply, I did not see the mention before somehow. I think this fix is okay for now to unblock people, but it would be great to have a configurable `CurlMaxTimeout`, since this change will add 30 seconds (6*5 additional seconds) to the test time.

Unfortunately, the only way `failRequest` should actually fail is due to a timeout, so it will wait 8 seconds each invocation.

qmonnet replied:
It will even add 6×8 seconds, since the 3-second long timeout was skipped so far (command not run). Yeah I can probably work out something to pass a configurable value. Could you please have a look at the other commit that changes some of the `failRequests()` to `doRequests()`?

gandro replied:
>  Could you please have a look at the other commit that changes some of the failRequests() to doRequests()?

I assume you are referring to https://github.com/cilium/cilium/pull/11307/commits/0be2205493bbbf0d261cc7a5abb1d86c9f1c9176 ?

If kube-proxy indeed forwards requests from local pods, then I think our implementation in NodePort BPF in the datapath is actually not compliant. The tests were written with NodePort BPF's implementation in mind, not kube-proxy. But I haven't worked on that code in a while, @brb would likely know more.

Edit: Discussed offline, our implementation should be compliant on a conceptual level. There might be a bug however.

Edit 2: After closer inspection, our implementation is incompatible with kube-proxy. kube-proxy forwards local requests, while cilium does not.

qmonnet replied:
> I think this fix is okay for now to unblock people, but it would be great to have a configurable `CurlMaxTimeout`, since this change will add 30 seconds (6*5 additional seconds) to the test time.
> 
> Unfortunately, the only way `failRequest` should actually fail is due to a timeout, so it will wait 8 seconds each invocation.

I had another look and this is in fact not the case, because we should timeout on the value passed to `--connect-timeout` (5 seconds) here and not the one from `--max-time`. Test logs confirm that. So no change needed on that side.

joestringer commented on test/helpers/kubectl.go:
```diff
@@ -2791,14 +2791,14 @@ func (kub *Kubectl) DumpCiliumCommandOutput(ctx context.Context, namespace strin
 
 			archiveName := filepath.Join(logsPath, fmt.Sprintf("bugtool-%s", pod))
-			res = kub.ExecContext(ctx, fmt.Sprintf("mkdir -p %s", archiveName))
+			res = kub.ExecContext(ctx, fmt.Sprintf("mkdir -p \"%s\"", archiveName))
```
Drive-by nit: Can/should we use `%q` for this purpose?

qmonnet replied:
What for? As I understand it, it [does not escape parenthesis](https://stackoverflow.com/questions/56055300/what-does-safely-escaped-with-go-syntax-mean) or enquote its argument, does it?

gandro replied:
While it does not escape all the characters (we would need something like https://godoc.org/github.com/alessio/shellescape for that), `%q` it would at least escape quotes, newline and other stuff which might break the command.

qmonnet replied:
Ah right and I was mistaken, the `%q` does seem to enquote the string. I'll use it then, thanks.

gandro reviewed (approved):
The new test setup for externalTrafficPolicy=Local looks good to me, thanks a lot for taking care of this!

qmonnet committed:
test/K8sServices: fix expectations in testExternalTrafficPolicyLocal()

The function testExternalTrafficPolicyLocal() used to check the
"externalTrafficPolicy=Local" flag for kube-proxy or its BPF replacement
uses a set of curl requests to check connectivity. Requests from k8s2 to
the test service (deployed on k8s2 with local policy) via k8s2 are
expected to succeed, while requests sent to k8s1 are all supposed to
fail.

However, it seems that this is not the expected behaviour. External
requests sent to k8s1 should fail indeed, but local requests, emitted
from pods located on k8s1, should not be affected and should be
forwarded to the backend on k8s2. This is visible in the iptables-save
dump from the cilium pod on k8s1 (extract):

    -A KUBE-XLB-FXQVMYTRN72EUJRL -s 10.10.0.0/16 -m comment --comment "Redirect pods trying to reach external loadbalancer VIP to clusterIP" -j KUBE-SVC-FXQVMYTRN72EUJRL
    -A KUBE-XLB-FXQVMYTRN72EUJRL -m comment --comment "masquerade LOCAL traffic for default/test-lb-local-k8s2:http LB IP" -m addrtype --src-type LOCAL -j KUBE-MARK-MASQ
    -A KUBE-XLB-FXQVMYTRN72EUJRL -m comment --comment "route LOCAL traffic for default/test-lb-local-k8s2:http LB IP to service chain" -m addrtype --src-type LOCAL -j KUBE-SVC-FXQVMYTRN72EUJRL
    -A KUBE-XLB-FXQVMYTRN72EUJRL -m comment --comment "default/test-lb-local-k8s2:http has no local endpoints" -j KUBE-MARK-DROP
    -A KUBE-XLB-R5LOCOWLR542QA7J -s 10.10.0.0/16 -m comment --comment "Redirect pods trying to reach external loadbalancer VIP to clusterIP" -j KUBE-SVC-R5LOCOWLR542QA7J
    -A KUBE-XLB-R5LOCOWLR542QA7J -m comment --comment "masquerade LOCAL traffic for default/test-nodeport-local-k8s2:http LB IP" -m addrtype --src-type LOCAL -j KUBE-MARK-MASQ
    -A KUBE-XLB-R5LOCOWLR542QA7J -m comment --comment "route LOCAL traffic for default/test-nodeport-local-k8s2:http LB IP to service chain" -m addrtype --src-type LOCAL -j KUBE-SVC-R5LOCOWLR542QA7J
    -A KUBE-XLB-R5LOCOWLR542QA7J -m comment --comment "default/test-nodeport-local-k8s2:http has no local endpoints" -j KUBE-MARK-DROP

Note the KUBE-MARK-DROP target is only used when the source is not of
"--src-type LOCAL". By contrast, k8s2 is able to route all requests to
its backend for the service:

    -A KUBE-XLB-FXQVMYTRN72EUJRL -s 10.10.0.0/16 -m comment --comment "Redirect pods trying to reach external loadbalancer VIP to clusterIP" -j KUBE-SVC-FXQVMYTRN72EUJRL
    -A KUBE-XLB-FXQVMYTRN72EUJRL -m comment --comment "masquerade LOCAL traffic for default/test-lb-local-k8s2:http LB IP" -m addrtype --src-type LOCAL -j KUBE-MARK-MASQ
    -A KUBE-XLB-FXQVMYTRN72EUJRL -m comment --comment "route LOCAL traffic for default/test-lb-local-k8s2:http LB IP to service chain" -m addrtype --src-type LOCAL -j KUBE-SVC-FXQVMYTRN72EUJRL
    -A KUBE-XLB-FXQVMYTRN72EUJRL -m comment --comment "Balancing rule 0 for default/test-lb-local-k8s2:http" -j KUBE-SEP-TJ5BBRFPOOF55I4D
    -A KUBE-XLB-R5LOCOWLR542QA7J -s 10.10.0.0/16 -m comment --comment "Redirect pods trying to reach external loadbalancer VIP to clusterIP" -j KUBE-SVC-R5LOCOWLR542QA7J
    -A KUBE-XLB-R5LOCOWLR542QA7J -m comment --comment "masquerade LOCAL traffic for default/test-nodeport-local-k8s2:http LB IP" -m addrtype --src-type LOCAL -j KUBE-MARK-MASQ
    -A KUBE-XLB-R5LOCOWLR542QA7J -m comment --comment "route LOCAL traffic for default/test-nodeport-local-k8s2:http LB IP to service chain" -m addrtype --src-type LOCAL -j KUBE-SVC-R5LOCOWLR542QA7J
    -A KUBE-XLB-R5LOCOWLR542QA7J -m comment --comment "Balancing rule 0 for default/test-nodeport-local-k8s2:http" -j KUBE-SEP-TZWVS7R6L6CPM73Z

But then, why did the CI fail to catch this before? This is because the
testCurlFailFromPodInHostNetNS() function used by the test does not call
helpers.CurlFail() correctly, and bash fails to parse and run the
command line, thus always returning an error code which we wrongly
interpreted as a failed attempt to connect. The request will be fixed in
a subsequent commit. For now, fix the test logics by expecting requests
from k8s1 to k8s1 to succeed.

Fixes: 54304819a1d2 ("test: Extend externalTrafficPolicy=Local tests")
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
test: enquote bugtool archive folder name in bash command

There are a couple of tests that contain parenthesis in their names.
Namely, in K8sServices:

- Tests NodePort (kube-proxy)
- Tests NodePort (kube-proxy) with externalTrafficPolicy=Local

Because of the parenthesis, handling directories named after the tests
can be more delicate in bash, and we need to make sure that the
parenthesis are not interpreted. So far this was not done properly, and
bash complains with:

    bash: -c: line 0: syntax error near unexpected token `('

Fix it by using the %q format specifier for the relevant bash commands.

Fixes: #11094
Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
test/K8sServices: skip checks for BPF with externalTrafficPolicy=Local

A previous commit fixed the expectations for the results for the curl
requests in testExternalTrafficPolicyLocal(). This should now be working
with kube-proxy, however it turns out the behaviour of the BPF
replacement for kube-proxy behaves differently for local traffic.

The test is still not detecting this issue because of the broken curl
requests that will be fixed in a subsequent patch, but let's skip and
comment the particular checks that are known to fail with Nodeport BPF.

Signed-off-by: Quentin Monnet <quentin@isovalent.com>

qmonnet committed:
test/K8sServices: fix a call to helpers.CurlFail()

helpers.CurlFail() can take optional arguments, and at one location this
is used to pass the "--max-time" option to the curl command line.

But these optional arguments for helpers.CurlFail() are not intended to
be used that way; they should instead be used to format the target URL,
which is supposed to be a string containing the relevant format
specifiers. Because in the current case, the URL provided has no format
specifier, go appends the additional arguments at the end, with
extraneous characters. So instead of producing the final curl command:

    curl [...] http://192.168.36.11:30861 --max-time 3 [...]

We actually get:

    curl [...] --max-time 20 http://192.168.36.11:30861%!(EXTRA string=--max-time 3) [...]

The first "--max-time 20" is added by helpers.CurlFail() itself.

Then of course bash fails to interpret that command, and complains with:

    bash: -c: line 0: syntax error near unexpected token `('

In order to fix it, let's remove the "--max-time 3" argument entirely.
As we can see from the resulting command line, "--max-time" is already
used (with "CurlMaxTimeout"). And it is useless anyway, since we
actually expect to timeout on the value passed through the
"--connect-timeout" option (currently 5 seconds).

While at it, let's also fix the typo in the comments for
helpers.CurlFail(), and remove the explicit argument indices used for
generating the curl command. They were useful at a time when one of the
arguments was used multiple times, but they are unnecessary and only
confusing now.

Fixes: 54304819a1d2 ("test: Extend externalTrafficPolicy=Local tests")
Signed-off-by: Quentin Monnet <quentin@isovalent.com>
